### ISSUE TITLE
Panic with VFIO with MSI only device (regression)

### ISSUE THREAD

**Issue by mehrdad-yousefi** [cloud-hypervisor/cloud-hypervisor#5649]
**Describe the bug**
I'm trying to start a VM and passthrough a NVIDIA Quadro T400 GPU. It shows me this error:

```
cloud-hypervisor: 124.120767ms: <vmm> ERROR:/root/.cargo/git/checkouts/vfio-3ea6ca0c73c7450b/89f8e77/crates/vfio-ioctls/src/vfio_device.rs:786 -- Could not get region #8 info
thread 'vmm' panicked at 'called `Option::unwrap()` on a `None` value', pci/src/vfio.rs:508:49
                                                                                              note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
                                                                                                                                                                           Failed to join on VMM thread: Any { .. }
```

When I look at `pci/src/vfio.rs:508`, it seems that it tries to access MSI-X, but it cannot because it's a `None` value. This is the output of `lspci`:

```
01:00.0 VGA compatible controller: NVIDIA Corporation TU117GLM [Quadro T400 Mobile] (rev a1) (prog-if 00 [VGA controller])
        Subsystem: NVIDIA Corporation TU117GLM [Quadro T400 Mobile]
        Control: I/O+ Mem+ BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Interrupt: pin A routed to IRQ 16
        IOMMU group: 1
        Region 0: Memory at a4000000 (32-bit, non-prefetchable) [size=16M]
        Region 1: Memory at 90000000 (64-bit, prefetchable) [size=256M]
        Region 3: Memory at a0000000 (64-bit, prefetchable) [size=32M]
        Region 5: I/O ports at 4000 [size=128]
        Expansion ROM at a5000000 [disabled] [size=512K]
        Capabilities: [60] Power Management version 3
                Flags: PMEClk- DSI- D1- D2- AuxCurrent=375mA PME(D0+,D1-,D2-,D3hot+,D3cold+)
                Status: D3 NoSoftRst+ PME-Enable- DSel=0 DScale=0 PME-
        Capabilities: [68] MSI: Enable- Count=1/1 Maskable- 64bit+
                Address: 0000000000000000  Data: 0000
        Capabilities: [78] Express (v2) Legacy Endpoint, MSI 00
                DevCap: MaxPayload 256 bytes, PhantFunc 0, Latency L0s unlimited, L1 <64us
                        ExtTag+ AttnBtn- AttnInd- PwrInd- RBE+ FLReset+
                DevCtl: CorrErr- NonFatalErr- FatalErr- UnsupReq-
                        RlxdOrd+ ExtTag+ PhantFunc- AuxPwr- NoSnoop+ FLReset-
                        MaxPayload 256 bytes, MaxReadReq 512 bytes
                DevSta: CorrErr- NonFatalErr- FatalErr- UnsupReq- AuxPwr+ TransPend-
                LnkCap: Port #0, Speed 8GT/s, Width x16, ASPM L0s L1, Exit Latency L0s <1us, L1 <4us
                        ClockPM+ Surprise- LLActRep- BwNot- ASPMOptComp+
                LnkCtl: ASPM Disabled; RCB 64 bytes, Disabled- CommClk+
                        ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
                LnkSta: Speed 8GT/s (ok), Width x16 (ok)
                        TrErr- Train- SlotClk+ DLActive- BWMgmt- ABWMgmt-
                DevCap2: Completion Timeout: Range AB, TimeoutDis+ NROPrPrP- LTR+
                         10BitTagComp- 10BitTagReq- OBFF Via message, ExtFmt- EETLPPrefix-
                         EmergencyPowerReduction Not Supported, EmergencyPowerReductionInit-
                         FRS-
                         AtomicOpsCap: 32bit- 64bit- 128bitCAS-
                DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis- LTR+ OBFF Disabled,
                         AtomicOpsCtl: ReqEn-
                LnkCap2: Supported Link Speeds: 2.5-8GT/s, Crosslink- Retimer- 2Retimers- DRS-
                LnkCtl2: Target Link Speed: 8GT/s, EnterCompliance- SpeedDis-
                         Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
                         Compliance De-emphasis: -6dB
                LnkSta2: Current De-emphasis Level: -3.5dB, EqualizationComplete+ EqualizationPhase1+
                         EqualizationPhase2+ EqualizationPhase3+ LinkEqualizationRequest-
                         Retimer- 2Retimers- CrosslinkRes: unsupported
        Capabilities: [100 v1] Virtual Channel
                Caps:   LPEVC=0 RefClk=100ns PATEntryBits=1
                Arb:    Fixed- WRR32- WRR64- WRR128-
                Ctrl:   ArbSelect=Fixed
                Status: InProgress-
                VC0:    Caps:   PATOffset=00 MaxTimeSlots=1 RejSnoopTrans-
                        Arb:    Fixed- WRR32- WRR64- WRR128- TWRR128- WRR256-
                        Ctrl:   Enable+ ID=0 ArbSelect=Fixed TC/VC=ff
                        Status: NegoPending- InProgress-
        Capabilities: [250 v1] Latency Tolerance Reporting
                Max snoop latency: 71680ns
                Max no snoop latency: 71680ns
        Capabilities: [258 v1] L1 PM Substates
                L1SubCap: PCI-PM_L1.2+ PCI-PM_L1.1+ ASPM_L1.2+ ASPM_L1.1+ L1_PM_Substates+
                          PortCommonModeRestoreTime=255us PortTPowerOnTime=10us
                L1SubCtl1: PCI-PM_L1.2- PCI-PM_L1.1- ASPM_L1.2- ASPM_L1.1-
                           T_CommonMode=0us LTR1.2_Threshold=0ns
                L1SubCtl2: T_PwrOn=10us
        Capabilities: [128 v1] Power Budgeting <?>
        Capabilities: [420 v2] Advanced Error Reporting
                UESta:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UEMsk:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UESvrt: DLP+ SDES+ TLP- FCP+ CmpltTO- CmpltAbrt- UnxCmplt- RxOF+ MalfTLP+ ECRC- UnsupReq- ACSViol-
                CESta:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr+
                CEMsk:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr+
                AERCap: First Error Pointer: 00, ECRCGenCap- ECRCGenEn- ECRCChkCap- ECRCChkEn-
                        MultHdrRecCap- MultHdrRecEn- TLPPfxPres- HdrLogCap-
                HeaderLog: 00000000 00000000 00000000 00000000
        Capabilities: [600 v1] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
        Capabilities: [900 v1] Secondary PCI Express
                LnkCtl3: LnkEquIntrruptEn- PerformEqu-
                LaneErrStat: 0
        Capabilities: [bb0 v1] Physical Resizable BAR
                BAR 0: current size: 16MB, supported: 16MB
                BAR 1: current size: 256MB, supported: 64MB 128MB 256MB
                BAR 3: current size: 32MB, supported: 32MB
        Kernel driver in use: vfio-pci
        Kernel modules: nouveau
```

You see that MSI-X capability is not there. Is there any workaround to get it work? I mean since my GPU doesn't have MSI-X capability, can I force `cloud-hypervisor` to use MSI capability instead? Or whatever change that might get the VM booted with the GPU passthrough. I had a successful PCIe passthrough for my Coral TPU device. So, I'm pretty sure, my setup should work fine.

**Version**

Output of `cloud-hypervisor --version`: `cloud-hypervisor v33.0-85-g0149e650`

Did you build from source, if so build command line (e.g. features): `bash -c "git clone https://github.com/cloud-hypervisor/cloud-hypervisor.git && cd cloud-hypervisor && /root/.cargo/bin/cargo build --release && setcap cap_net_admin+ep ./target/release/cloud-hypervisor && /root/.cargo/bin/cargo build --release --target=x86_64-unknown-linux-musl"`

**VM configuration**

What command line did you run (or JSON config data):

Guest OS version details: Ubuntu 22.04

Host OS version details: CentOS 8

**Logs**

Output of `cloud-hypervisor -v` from either standard error or via `--log-file`:

Linux kernel output:

**Comment by rbradford** [comment]
I would boot a modern distro image (using OVMF or Rust Hypervisor Firmware) and use the distro support for the nvidia proprietary drivers. Kernel 5.4 seems extremely old at this point.

**Comment by rbradford** [comment]
@mehrdad-yousefi Fix in #5658 if you don't mind trying it.
